Automattic/wp-calypso
# Onboarding: Update the FSE Beta flow

#### Changes proposed in this Pull Request

* Move the FSE Beta opt-in step in first position for users eligible for the FSE Beta.
* Update the FSE Beta opt-in step design.

| Before | After |
| - | - |
| <img width="1437" alt="Screenshot 2021-10-13 at 12 49 43" src="https://user-images.githubusercontent.com/2070010/137128044-287c6f93-8f7c-4db2-9856-9235f146e67c.png"> | <img width="948" alt="Screenshot 2021-10-19 at 17 55 04" src="https://user-images.githubusercontent.com/2070010/137971675-71d4678d-8476-4603-94e9-ffe254c06063.png"> |

#### The New FSE Beta Flow

As discussed in paYE8P-13J-p2, we should move the FSE Beta opt-in step in first position in the onboarding flow.
As far as I can see, Gutenboarding was designed to have the `IntentGathering` step always in first position, among other reasons because it's well integrated in the whole routing logic.
In a [previous attempt](https://github.com/Automattic/wp-calypso/pull/57155), I tried to conditionally render the FSE Beta opt-in step _inside_ `IntentGathering`, but eventually I realized that it would have messed with the browser's history more than I'd deem reasonable.

This approach, instead, takes advantage of the routing setup and, **for users eligible for the FSE Beta**: 

1. Redirects the root path (`/`) to the FSE Beta opt-in step (`/beta`).
2. Renders `IntentGathering` in a new `/title` path (instead of `/` which now gets redirected).
3. Adds `/beta` and `/title` as first and second steps of the FSE Beta flow.
4. Gives a back button to `IntentGathering` now that it's the second step.

#### Why Fixing Two PRs at Once?

This PR fixes both the issue for updating the FSE Beta opt-in step design, and for moving it in first position.
The reason is that the new design was created to be in first position in the first place (for example, it does not have a back button).

#### Why a New FSE Beta Hook?

Here I have created a new `useFseBetaOptInStep` hook, in addition to the existing `useFseBetaEligibility`.

The new hook determines if a user can see the FSE Beta opt-in step.
From what I could gather by poking around Gutenboarding, this happens when:
- The user is eligible for the FSE Beta (of course).
- The onboarding is not using the Anchor integration flow.

I'm not familiar with the Anchor flow, but it's a simple check (that also happens pretty much everywhere in Gutenboarding) that couldn't hurt.

Now the old `useFseBetaEligibility` hook is only ever used inside the new one.
I've decided to keep them separate nonetheless in order to have them take care of very different concerns independently.

#### Testing instructions

* Open Gutenboarding at `http://calypso.localhost:3000/new`.
* Make sure you are automatically redirected to `/new/beta`.
* It shouldn't have created a new history stack item, so using the browser's back button should return to the empty tab (or whatever you have set up).
* Test the new design, and make sure it looks good on any screen size.
* Move around Gutenboarding, both with the on-screen buttons, and the browser's ones.
* Make sure it works as expected.

To test the regular flow, we need to manually turn off the FSE Beta feature flag.
* Open `http://calypso.localhost:3000/new?flags=-full-site-editing/beta-opt-in`.
* Make sure the first step is `IntentGathering` (the title screen) and that there hasn't been any redirects (the path is still `/new`).
* Move around Gutenboarding, both with the on-screen buttons, and the browser's ones.
* Make sure it works as expected.




Fixes #57092
Fixes #57093
